Backend/src/main/java/ca/mcgill/cooperator/controller/CourseController.java

```diff
@@ -17,18 +27,94 @@
 @RequestMapping("courses")
 public class CourseController {
 
-    @Autowired CourseService courseService;
+    @Autowired private CourseService courseService;
+    @Autowired private CourseOfferingService courseOfferingService;
 
+    /**
+     * Get a Course by ID
+     *
+     * @param id
+     * @return CourseDto object
+     */
     @GetMapping("/{id}")
     public CourseDto getCourseById(@PathVariable int id) {
+        Course c = courseService.getCourseById(id);
+        return ControllerUtils.convertToDto(c);
+    }
+
+    /**
+     * Get all courses
+     *
+     * @return List of CourseDto objects
+     */
+    @GetMapping("")
+    public List<CourseDto> getAllAdmins() {
+        List<Course> courses = courseService.getAllCourses();
 
-        return ControllerUtils.convertToDto(courseService.getCourseById(id));
+        return ControllerUtils.convertCourseListToDto(courses);
     }
 
+    /**
+     * Create a new Course
+     *
+     * <p>In request body:
+     *
+     * @param course name
+     */
     @PostMapping("")
-    public CourseDto createCourse(@RequestBody CourseDto courseDto) {
-        Course course = courseService.createCourse(courseDto.getName());
+    public CourseDto createCourse(@RequestBody CourseDto c) {
+        Course createdCourse = courseService.createCourse(c.getName());
+
+        return ControllerUtils.convertToDto(createdCourse);
+    }
+
+    /**
+     * Update an existing Course
+     *
+     * <p>In request body:
+     *
+     * @param name
+     */
+    @PutMapping("")
+    public CourseDto updateCourse(@RequestBody CourseDto c) {
+        Course course = courseService.getCourseById(c.getId());
+        Course updatedCourse =
+                courseService.updateCourse(
+                        course,
+                        c.getName(),
+                        convertCourseOfferingListToDomainObject(c.getCourseOfferings()));
+
+        return ControllerUtils.convertToDto(updatedCourse);
+    }
+
+    /**
+     * Delete an existing Course
+     *
+     * @param id
+     * @return deleted course
+     */
+    @DeleteMapping("/{id}")
+    public CourseDto deleteCourse(@PathVariable int id) {
+        Course course = courseService.getCourseById(id);
+        Course deletedCourse = courseService.deleteCourse(course);
+
+        return ControllerUtils.convertToDto(deletedCourse);
+    }
+
+    @ExceptionHandler(RuntimeException.class)
+    public final ResponseEntity<Exception> handleAllExceptions(RuntimeException ex) {
+        return new ResponseEntity<Exception>(ex, HttpStatus.INTERNAL_SERVER_ERROR);
+    }
 
-        return ControllerUtils.convertToDto(course);
+    private List<CourseOffering> convertCourseOfferingListToDomainObject(
```

Comment:
This should probably go in ControllerUtils?

Reply:
@kraglalbert i had it there but it cant be static so i moved it here (((i think that was the error but i cant remember exactly but i moved it out of controllerutils on purpose)))

Reply:
oh i think that it was that the course service needed in a method was null when i was testing even though it was autowired